server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -164,6 +175,10 @@ public String[] concreteIndexNames(ClusterState state, IndicesOptions options, I
         return concreteIndexNames(context, request.indices());
     }
 
+    public List<String> matchesSystemIndexPattern(String... indexExpressions) {
```

Comment:
@andrross I know you have suggested the change here but this API looks like one off to me: 

```
concreteIndexNames(ClusterState, IndicesOptions, boolean, String...)
concreteIndexNames(ClusterState, IndicesOptions, IndicesRequest)
concreteIndexNames(ClusterState, IndicesOptions, String...)
concreteIndexNames(ClusterState, IndicesRequest)
concreteIndexNamesWithSystemIndexAccess(ClusterState, IndicesRequest)
concreteIndices(ClusterState, IndicesOptions, boolean, String...)
concreteIndices(ClusterState, IndicesOptions, String...)
concreteIndices(ClusterState, IndicesRequest)
concreteIndices(ClusterState, IndicesRequest, long)
concreteSingleIndex(ClusterState, IndicesRequest)
concreteWriteIndex(ClusterState, IndicesOptions, String, boolean, boolean)
concreteWriteIndex(ClusterState, IndicesRequest)
dataStreamNames(ClusterState, IndicesOptions, String...)
```

I would suggest to align it with other APIs:

```
public List<String> systemIndexNames(ClusterState state, IndicesOptions options, String... indexExpressions)
```

My primary concern is that the system indices are picked from the static data sources (plugins) and do reflect the desired state but not actual state. For example, cluster may be in recovery state and not all system indices are available (or could be forceably closed)

WDYT?
```

Reply:
From the security plugin perspective, ideally I would like to call on `concreteIndexNames` once and then determine if any of those match any system index patterns.

1. This is where security resolves the indices in a request to concrete indices: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L311-L313
2. This is where security uses the resolved concrete indices and determines if any of them match system index patterns: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/SecurityIndexAccessEvaluator.java#L174-L180

The primary problem I'm trying to solve with this PR is to rely on the system index registry in core instead of relying on a [list](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/tools/democonfig/SecuritySettingsConfigurer.java#L49-L80) maintained in the security plugin

Reply:
> My primary concern is that the system indices are picked from the static data sources (plugins) and do reflect the desired state but not actual state.

Yeah, generally I would prefer to align this with the other APIs as you suggested.

System indexes are a bit different though, because I believe plugins define a pattern and then all current, future, and past indexes that match that pattern are system indexes. @cwperks What is the use case here? Does this call need to be scoped to currently-existing indexes?

Reply:
@reta @andrross I don't know if it will help with API design, but I opened a Draft PR that I was working on as a follow-up to this: https://github.com/opensearch-project/OpenSearch/pull/14630

What I am working towards is strong ownership of system indices by the plugins and preventing what plugins can do in a block where the threadcontext is stashed to only actions interacting with the plugin's system indices. 

That PR uses the proxy pattern to populate an ExecutionContext with the canonical class name of the plugin before execution is delegated to the plugin. The execution context is cleared on return.

I will also need another method to determine if the request is only scoped to system indices that are owned by the plugin that is currently executing.

Reply:
> System indexes are a bit different though, because I believe plugins define a pattern and then all current, future, and past indexes that match that pattern are system indexes.

Oh I see, would be `SystemIndexRegistry` (or alike) be better abstraction than? I think it will be matching the static nature of the system index patterns (much like `MediaTypeRegistry`, ....). For some reasons I was thinking about the case of resolving system indices to concrete indices but this is clearly not what is being done.

Reply:
@reta the [SytemIndices](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/SystemIndices.java) class is a registry

Reply:
@cwperks thanks, that should be the place than? As an idea, we could extract the `SERVER_SYSTEM_INDEX_DESCRIPTORS` from `SystemIndices` to `SystemIndexRegistry` since we would need to provide static access to it;

```
class SystemIndexRegistry {
 public static List<String> matchesSystemIndexPattern(String... indexExpressions)  {
 ...
 }
}
```

Reply:
@reta how can static access be provided? During bootstrap the core collects the SystemIndexPlugins and all of their descriptors to initialize the registry.

This is leading me back to the initial design of the PR where a single plugin can obtain the registry. Thoughts?

Reply:
> This is leading me back to the initial design of the PR where a single plugin can obtain the registry. Thoughts?

`SystemIndices` exposes a lot of things and is not supposed to be used in static context, whereas we could seal `SystemIndexRegistry` instantiation and only have static accessor(s).

Reply:
ok, pushed a change to create a `SystemIndexRegistry` that is a singleton. Its initialized once in the constructor of SystemIndices which itself is instantiated in [Node.java](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/node/Node.java#L829)

For testing purposes, I had to create a method called `clear()` to null-ify the instance before test run.